drivers: usdhc: Fix bug with multiple writes failing

danieldegrasse opened the pull request:
The NXP USDHC driver will fail to write to an SD card if `disk_access_write` is called multiple times without a delay. This is due to the USDHC driver neglecting to wait for the SD card to be idle before attempting to send it new commands. 

This PR fixes that bug, and adds a test for disk drivers, in order to verify that multiple reads and writes work correctly.

Fixes #39942

danieldegrasse committed:
tests: drivers: disk: Add disk driver test

Adds a disk driver test, intended for the NXP USDHC driver but valid for
any disk device. The test performs multiple reads and writes to the
disk, and verifies that the driver does not write out of bounds to the
provided buffer.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: disk: USDHC: Update USDHC driver to pass disk test

USDHC driver did not pass disk test (failed during multiple writes). Add
logic to wait for the SD card to be idle before reading new data, so
that the test will pass. Also, add logic to reject OOB reads and writes
to the disk.

Fixes #39942

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

jfischer-no commented on tests/drivers/disk/src/main.c:
```diff
@@ -0,0 +1,219 @@
+#include <storage/disk_access.h>
+#include <device.h>
+
+#if defined(CONFIG_SDMMC_USDHC)
```
Please change it to:
```diff
diff --git a/tests/drivers/disk/src/main.c b/tests/drivers/disk/src/main.c
index b276b649085..f9bcbb7b99e 100644
--- a/tests/drivers/disk/src/main.c
+++ b/tests/drivers/disk/src/main.c
@@ -14,8 +14,10 @@
 #include <storage/disk_access.h>
 #include <device.h>
 
-#if defined(CONFIG_SDMMC_USDHC)
-#define DISK_NAME "SDMMC"
+#if IS_ENABLED(CONFIG_DISK_DRIVER_SDMMC)
+#define DISK_NAME CONFIG_SDMMC_VOLUME_NAME
+#elif IS_ENABLED(CONFIG_DISK_DRIVER_RAM)
+#define DISK_NAME CONFIG_DISK_RAM_VOLUME_NAME
 #else
 #error "No disk device defined, is your board supported?"
 #endif

```

regarding CONFIG_DISK_DRIVER_RAM, I tested your PR on top of https://github.com/zephyrproject-rtos/zephyr/pull/40611

jfischer-no commented on tests/drivers/disk/prj.conf:
```diff
@@ -0,0 +1,3 @@
+CONFIG_TEST=y
+CONFIG_ZTEST=y
+CONFIG_DISK_ACCESS=y
```
`+ CONFIG_DISK_DRIVER_SDMMC=y`

danieldegrasse replied:
Should `CONFIG_DISK_DRIVER_SDMMC` be selected unconditionally, even when testing a RAM disk?

jfischer-no replied:
I think one should be enabled per default, you PR is targeting SDMMC driver, I can add test case for ramdisk in #40611

danieldegrasse replied:
Sounds good. `CONFIG_DISK_DRIVER_SDMMC` should be selected for the test in the latest push

jfischer-no commented on tests/drivers/disk/src/main.c:
```diff
@@ -0,0 +1,219 @@
+
+	/* Verify that multiple writes to the same location work */
+	i = 10;
+	while (i--) {
```
`for (int i = 0; i < 10; i++) { ` ?

jfischer-no commented on tests/drivers/disk/src/main.c:
```diff
@@ -0,0 +1,219 @@
+	rc = read_sector(scratch_buf[0], 0, SECTOR_COUNT1);
+	zassert_equal(rc, 0, "Failed to read from disk");
+	i = 10;
+	while (i--) {
```
`for (int i = 0; i < 10; i++) { ` ?

jfischer-no commented on tests/drivers/disk/src/main.c:
```diff
@@ -0,0 +1,219 @@
+		memset(scratch_buf[1], 0xff, SECTOR_COUNT1 * disk_sector_size);
+		rc = read_sector(scratch_buf[1], 0, SECTOR_COUNT1);
+		zassert_equal(rc, 0, "Failed to read from disk at same sector location");
+		zassert_equal(memcmp(scratch_buf[1], scratch_buf[0],
```
```diff
diff --git a/tests/drivers/disk/src/main.c b/tests/drivers/disk/src/main.c
index b276b649085..c76af9cde5d 100644
--- a/tests/drivers/disk/src/main.c
+++ b/tests/drivers/disk/src/main.c
@@ -178,9 +178,9 @@ static void test_read(void)
                memset(scratch_buf[1], 0xff, SECTOR_COUNT1 * disk_sector_size);
                rc = read_sector(scratch_buf[1], 0, SECTOR_COUNT1);
                zassert_equal(rc, 0, "Failed to read from disk at same sector location");
-               zassert_equal(memcmp(scratch_buf[1], scratch_buf[0],
-                               SECTOR_COUNT1 * disk_sector_size), 0,
-                               "Multiple reads from same disk location did not return same data");
+               zassert_mem_equal(scratch_buf[1], scratch_buf[0],
+                                 SECTOR_COUNT1 * disk_sector_size,
+                                 "Multiple reads mismatch");
        }
 }
```

Please keep it short.

jfischer-no replied:
you can use `zassert_mem_equal` instead of zassert_equal(memcmp())

```
+               zassert_mem_equal(scratch_buf[1], scratch_buf[0],
+                                 SECTOR_COUNT1 * disk_sector_size,
+                                 "Multiple reads mismatch");
```

danieldegrasse replied:
fixed, sorry I missed that change before.

jfischer-no commented on tests/drivers/disk/testcase.yaml:
```diff
@@ -0,0 +1,8 @@
+tests:
+  drivers.disk.usdhc:
```
I guess `platform_allow` can be replaced by `filter: CONFIG_SDMMC_USDHC`

with https://github.com/zephyrproject-rtos/zephyr/pull/40611, you can add a test case for any platform:
```
  drivers.disk.ram:
    min_ram: 128
    tags: disk
    extra_configs:
        - CONFIG_DISK_DRIVER_RAM=y
...
```

danieldegrasse replied:
Added this test case. I'm assuming #40611 is required to make this test pass, since it checks for out of bounds r/w?

danieldegrasse replied:
thanks for the heads up, updated the testcase.yaml file to use `filter: CONFIG_SDMMC_USDHC`

danieldegrasse committed:
soc: mimxrt685_evk: Fix usdhc driver build failure

USDHC driver build failure fix for mimxrt685_evk, dat3 pull function was
not defined at the SOC level.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>
